### ISSUE TITLE
[Question] How do I type object that is extended after variable declaration?

### ISSUE THREAD

**Issue by ajcpwnz** [microsoft/TypeScript#36258]
**TypeScript Version:**  3.7.4
**Search Terms:**
How do I type object that is extended after variable declaration?
**Code**
Part of a snippet generated by sequelize, with simple types added by me, where models are loaded and assigned into db object.
line `const db: DBIndexShape = {};` triggers warning for missing keys, even though they are assigned later within the loop. 

What are ways of dealing with this warning except:
- @ts-ignore
- creating `db` within reduce loop with type declaration 
- making keys optional?

```ts
interface ModelIndex {
    User: Model
}
interface Model {
    name: keyof ModelIndex
    associate?: (db: DBIndexShape) => void
}

interface DBIndexShape extends ModelIndex {
    sequelize: typeof sequelize
    Sequelize: typeof Sequelize
}

const db: DBIndexShape = {};

fs
    .readdirSync(__dirname)
    .filter((file: string) => {
        return (file.indexOf('.') !== 0) && (file !== basename) && (file.slice(-3) === (process.env.NODE_ENV === 'development' ? '.ts' : '.js'));
    })
    .forEach((file: string) => {
        const model: Model = sequelize['import'](path.join(__dirname, file));
        db[model.name] = model;
    });

```

**Expected behavior:**
Some conventional way to type this

**Actual behavior:**
None that I'm aware of 

**Playground Link:** 
not required

**Related Issues:**
none

**Comment by orta** [comment]
Hey there, this looks like a great example of a question that you should ask on [StackOverflow](http://stackoverflow.com) - there is an active [TypeScript](http://stackoverflow.com/tags/typescript) tag where people can help out with questions like this. We try to keep the TypeScript issues focused only on bug reports for the language, feature request for the type system. Which this doesn't look like.
